Migrate `AsyncContext` to `ContextMap` API

idelpivnitskiy opened the pull request:
Motivation:

#1904 introduced a new `ContextMap` API, which will handle context for
all use-cases.

API modifications:

- Deprecate `AsyncContextMap` in favor of `ContextMap` API;
- Deprecate `AsyncContextMapHolder` in favor of `ContextMapHolder`;
- `AsyncContext`: deprecate all methods that reference `AsyncContextMap`
and add new static methods that target `ContextMap` API;
- Add temporarily method `AsyncContext.newKeyMapping` to allow mapping
`AsyncContextMap.Key` <-> `ContextMap.Key` to let users use old/new API
at the same codebase;
- `IoThreadFactory.IoThread` extends deprecated `AsyncContextMapHolder`,
consider migrating to `NettyIoExecutors` static factories that do not
require an `IoThreadFactory`. If your codebase implements
`IoThreadFactory`, make sure it also implements `ContextMapHolder`;
- `BasicAuthHttpServiceFilter`: deprecate
`userInfoKey(AsyncContextMap.Key)` in favor of
`userInfoAsyncContextKey(ContextMap.Key)` and
`userInfoRequestContextKey(ContextMap.Key)`;
- `BasicAuthSecurityContextFilters`: deprecate
`forGlobalBinding(AsyncContextMap.Key)` in favor of
`forGlobalBinding(ContextMap.Key)` and
`forNameBinding(AsyncContextMap.Key)` in favor of
`forNameBinding(ContextMap.Key)`;

Behavior changes:

- `AsyncContext.put(...)` and `AsyncContext.putAll(...)` don't throw
`UnsupportedOperationException` when `AsyncContext` is disabled;
- Default implementation of `AsyncContext` more strictly validates
`null` keys. `null` keys were never allowed, but pre-existing behavior
throws NPE on `AsyncContext.get(key)` instead of
`AsyncContext.put(null, value)`;

Internal modifications:

- Migrate all existing use of `AsyncContextMap` to `ContextMap`;
- Update `CopyOnWriteAsyncContextMap` and `ConcurrentAsyncContextMap` to
target new `ContextMap` API;
- Add tests to verify users of `AsyncContextMap` API are not broken;
- Add `AsyncContextMapToContextMapAdapter`;
- Add tests to verify `AsyncContextMap.Key` and `ContextMap.Key` can be
used simultaneously if it was registered using
`AsyncContext.newKeyMapping`;
- Update links in `async-context.adoc`;
- Log when `AsyncContext` is enabled or disabled;

Result:

Unified context API.

bondolo commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/ConcurrentAsyncContextMap.java:
```diff
@@ -15,30 +15,37 @@
 import javax.annotation.Nullable;
 
-import static io.servicetalk.concurrent.api.AsyncContextMapUtils.contextMapToString;
+final class ConcurrentAsyncContextMap implements ContextMap {
```
It doesn't always work but one way to work around this is to branch off of main, `git mv` the file(s) to their new names but make no other changes, commit and and then rebase on top of that change. Sometimes this will convince it to track the name change and other times it will still just forget.

idelpivnitskiy replied:
Correct, `git mv` helps when you have two independent commits/PRs. Inside one PR (and after "squash & merge") it will be again shown as a new file. I tried to avoid that.

idelpivnitskiy replied:
Renamed in #1959

bondolo commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/CopyOnWriteAsyncContextMap.java:
```diff
@@ -15,23 +15,31 @@
+ * user entries).
  */
-final class CopyOnWriteAsyncContextMap implements AsyncContextMap {
+final class CopyOnWriteAsyncContextMap implements ContextMap {
```
I do think that long term we do want the new names. Since the classes are both package private the renaming does seem safe.

idelpivnitskiy replied:
Agreed, I can rename them in a follow-up or right before merging (after review is complete).

idelpivnitskiy replied:
Renamed in ##1959

bondolo commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AsyncContextProvider.java:
```diff
@@ -40,200 +41,214 @@
-     * @param newContextMap The current context.
-     */
-    void contextMap(AsyncContextMap newContextMap);
+    ContextMap context();
```
Missing Javadoc

bondolo commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AsyncContextProvider.java:
```diff
@@ -40,200 +41,214 @@
      * @return The wrapped {@link Callable}.
      */
-    <V> Callable<V> wrapCallable(Callable<V> callable, AsyncContextMap contextMap);
+    <V> Callable<V> wrapCallable(Callable<V> callable, ContextMap current);
```
The naming should be more reflective of the use, perhaps `context`?

bondolo commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AsyncContextProvider.java:
```diff
@@ -40,200 +41,214 @@
+    Runnable wrapRunnable(Runnable runnable, ContextMap current);
 
     /**
      * Wrap a {@link Callable} to ensure it is able to track {@link AsyncContext} correctly.
```
Perhaps change to `ContextMap` as well?

idelpivnitskiy replied:
This is a helper pkg-private interface to support `AsyncContext`, referencing `AsyncContext` from here is intentional.

bondolo commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/NoopAsyncContextProvider.java:
```diff
@@ -116,48 +121,50 @@ public ExecutorService wrapJdkExecutorService(final ExecutorService executor) {
 
     @Override
-    public Runnable wrapRunnable(final Runnable runnable, final AsyncContextMap contextMap) {
+    public Runnable wrapRunnable(final Runnable runnable, final ContextMap current) {
```
I'd still prefer `context` to `current`. Using current almost implies we shouldn't bother to ask for the context and just use the `AsyncContext.context()` current context.

idelpivnitskiy replied:
Agreed, it was a mix of current/contextMap in the original `AsyncContextProvider`. I aligned them to `current`, but will rename to `context` as you suggested.

bondolo commented on servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/IoThreadFactory.java:
```diff
@@ -32,8 +32,8 @@
      * interface.
      */
+    // FIXME: 0.42 - replace AsyncContextMapHolder with ContextMapHolder
     interface IoThread extends AsyncContextMapHolder {
```
Perhaps have it extend both?

idelpivnitskiy replied:
If I add `ContextMapHolder` now, I have to provide some default implementations. Correct implementation requires a class variable, which is not supported by interfaces. 
Throwing `UnsupportedOperationException` from `ContextMapHolder` will be a breaking change because our concurrent-api code uses them.

idelpivnitskiy commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AsyncContext.java:
```diff
@@ -104,11 +226,30 @@ public static AsyncContextMap current() {
+     * implementation.
+     * @see ContextMap#putAll(Map)
+     */
+    public static void putAllFromMap(final Map<ContextMap.Key<?>, Object> map) {
```
Java complains if I have two methods that take a `Map` of something, because parameters will be replaced at compilation time. Therefore, had to come up with a different name. We can add a `putAll` alias (and deprecate `putAllFromMap`) in 0.42. I will add a comment to clarify.

idelpivnitskiy commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AsyncContext.java:
```diff
@@ -133,20 +294,38 @@ public static void putAll(Map<AsyncContextMap.Key<?>, Object> map) {
      */
-    public static boolean removeAll(Iterable<AsyncContextMap.Key<?>> entries) {
+    @Deprecated
+    public static boolean removeAll(final Iterable<AsyncContextMap.Key<?>> entries) {
```
Java complains if I have two methods that take a `Map` of something, because parameters will be replaced at compilation time. Therefore, had to come up with a different name. We can add a `putAll` alias (and deprecate `putAllFromMap`) in 0.42. I will add a comment to clarify.

idelpivnitskiy replied:
Added an alias in #1959

idelpivnitskiy commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AsyncContext.java:
```diff
@@ -175,19 +391,77 @@ public static void clear() {
+     *
+     * @param key The {@link ContextMap.Key} to lookup.
+     * @param value The value to match.
+     * @param <T> The anticipated type of object associated with the {@code key}.
```
This is the order in how Intellij IDEA generates javadoc. Looks consistent with all other places we have.

bondolo replied:
I've noticed that if you use the "Change signature" refactoring tools it resorts generic types to the beginning. Weird, perhaps inconsistent within IntelliJ IDEA.

idelpivnitskiy commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AsyncContext.java:
```diff
@@ -175,19 +391,77 @@ public static void clear() {
+     * @param <T> The anticipated type of object associated with the {@code key}.
+     * @return {@code true} if this context contains a key-value pair matching the passed {@code key} and
+     * {@code value}, {@code false} otherwise.
+     * @throws NullPointerException (optional behavior) if {@code key} or {@code value} is {@code null} and the
```
I'm keeping it consistent with pre-existing contract of `AsyncContextMap`. It defines only value as a `@Nullable`, but uses the same javadoc for `NullPointerException`. IIUC, `AsyncContextMap` tries to be as `Map`, which allows implementations that support `null` keys, but doesn't put an explicit `@Nullable` annotation because it doesn't make much sense for us.

idelpivnitskiy commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AsyncContext.java:
```diff
@@ -199,18 +473,35 @@ public static boolean isEmpty() {
+    /**
+     * Convenience method to iterate over the key-value pairs contained in the current context.
+     *
+     * @param consumer Each key/value pair will be passed as arguments to this {@link BiPredicate}. Returns {@code true}
```
The main goal of its implementation is to "consume" key-value pair. The fact that this is a predicate is just an enhancement to let users abort iteration.

idelpivnitskiy commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AsyncContext.java:
```diff
@@ -199,18 +473,35 @@ public static boolean isEmpty() {
+     * @see ContextMap#forEach(BiPredicate)
+     */
+    @Nullable
+    public static ContextMap.Key<?> forEachPair(final BiPredicate<ContextMap.Key<?>, Object> consumer) {
```
It doesn't signal matching or not-matching. It signals when to "break" from the look. Users who want to iterate through the whole map will always return `true`. This API replicates a [ByteProcessor](https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/ByteProcessor.java) pattern from netty. Pre-existed in `AsyncContextMap` API. Gives a bit more flexibility to the users without expanding API surface.

idelpivnitskiy commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AsyncContextMapThreadLocal.java:
```diff
@@ -15,46 +15,30 @@
+
 import static java.lang.ThreadLocal.withInitial;
 
 final class AsyncContextMapThreadLocal {
```
Added `FIXME` comment to remind myself

idelpivnitskiy replied:
Actually, it might be better to keep the name as-is, because the public API this class supports is `AsyncContext`.

idelpivnitskiy commented:
This PR was automatically closed after I merged #1904 😢 
I've reopened it, changed the target branch to `main`, and rebased. Please, continue reviewing.

bondolo commented on servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/NettyIoThreadFactory.java:
```diff
@@ -94,23 +100,52 @@ public String toString() {
+            if (ASYNC_CONTEXT_MAP_DETECTED.compareAndSet(false, true)) {
+                LOGGER.warn("Detected usage of deprecated {}, migrate your code to {}",
+                        AsyncContextMapHolder.class.getCanonicalName(), ContextMapHolder.class.getCanonicalName(),
+                        new RuntimeException("Stack trace where AsyncContextMapHolder was used"));
```
I have always preferred `Throwable` for stack dumps

bondolo reviewed (approved):
136/151 files, but I am satisfied based on the current state of PR and tests.

idelpivnitskiy commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AsyncContext.java:
```diff
@@ -68,11 +76,42 @@ static AsyncContextProvider provider() {
+     * {@link ContextMap} API. This method will be removed in future releases.
+     */
+    @Deprecated
+    public static <T> void newKeyMapping(final AsyncContextMap.Key<T> acmKey, final ContextMap.Key<T> cmKey) {
```
> Have you considered having the registration happen implicitly with a possible static utility to enable/disable the behaviour?

We give a chance to users to register a mapping upfront. If nothing was registered, we always have to do that implicitly. Disabling that implicit mapping will break old API users because internal implementation is already migrated to the new API.


> Registering a mapping at `Key` creation time might not be too much overhead

Hm... This sounds like a good idea performance-wise. However, if I register all mappings upfront implicitly (at `newKey` time) and let users override mappings through this public utility later, they will lose visibility into those `info` level logs in `AsyncContextMapToContextMapAdapter` that help to identify which keys still use deprecated API.  WDYT?

chemicL replied:
Got it. I actually made that remark before seeing the implicit registrations :) Regarding the optimization and info logs about deprecated API -> it's a tradeoff -> perhaps we should measure what performance hit will typical workloads take with and without the logs and decide based on that data.

idelpivnitskiy replied:
This mapping is required only for 0.41.x and will be removed in 0.42. We can prioritize optimization if someone reports a visible impact. In practice, users don't have too many keys created at runtime. The hit should happen only for the first use of a key.

idelpivnitskiy replied:
Removed in #1959

idelpivnitskiy commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AsyncContextMapToContextMapAdapter.java:
```diff
@@ -0,0 +1,175 @@
+            return;
+        }
+        final Map<ContextMap.Key<?>, Object> cMap = new HashMap<>(map.size());
+        for (Map.Entry<AsyncContextMap.Key<?>, Object> entry : map.entrySet()) {
```
`putAll` is an atomic operation in `CopyOnWriteAsyncContextMap`, used an intermediate map to keep the behavior "all or try again". I will leave a comment in the code.

idelpivnitskiy commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AsyncContextProvider.java:
```diff
@@ -40,200 +41,219 @@
      * @return The current context.
+     * @deprecated Use {@link #context()}.
      */
+    @Deprecated
```
It's not necessary but helps to find what exactly has to be removed through IDE search or compiler warnings

idelpivnitskiy commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/ConcurrentAsyncContextMap.java:
```diff
@@ -15,30 +15,38 @@
 import javax.annotation.Nullable;
 
-import static io.servicetalk.concurrent.api.AsyncContextMapUtils.contextMapToString;
+// FIXME: 0.42 - rename to ConcurrentContextMap
```
I kept old names to allow reviewers see the diff, but will rename before merging this PR. See https://github.com/apple/servicetalk/pull/1910#discussion_r735925117 for more context.

idelpivnitskiy commented on servicetalk-grpc-netty/src/main/java/io/servicetalk/grpc/netty/DefaultGrpcServerBuilder.java:
```diff
@@ -59,7 +59,7 @@
 import static io.servicetalk.concurrent.internal.FutureUtils.awaitResult;
 import static io.servicetalk.grpc.api.GrpcExecutionStrategies.defaultStrategy;
-import static io.servicetalk.grpc.internal.DeadlineUtils.GRPC_DEADLINE_KEY;
+import static io.servicetalk.grpc.internal.DeadlineUtils.GRPC_DEADLINE_CONTEXT_KEY;
```
Agreed. Do you have any suggestions for a better name for 0.41? Or I can keep it as-is for 0.41, but rename it back to the old `GRPC_DEADLINE_KEY` in a follow-up that will clean up all deprecations. Prefer to be extra cautious for 0.41 even for internal API.

idelpivnitskiy replied:
Renamed in #1959

tkountis commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AsyncContext.java:
```diff
@@ -175,19 +393,77 @@ public static void clear() {
+    }
+
+    /**
+     * Convenience method to determine the number of key/value pairs in the current context.
```
Consider "key/value pairs" -> "entries" in the whole class. Java users are familiar with the term.
Same for API, eg. `forEachEntry`.
A matter of taste so no strong feelings.

tkountis commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AsyncContext.java:
```diff
@@ -369,5 +664,6 @@ private static void enable0() {
     private static void disable0() {
         provider = NoopAsyncContextProvider.INSTANCE;
         EXECUTOR_PLUGINS.remove(EXECUTOR_PLUGIN);
+        LOGGER.info("Disabled, features that depend on AsyncContext will stop working.");
```
I think it should be a dot here rather than a comma.

tkountis commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AsyncContextMapToContextMapAdapter.java:
```diff
@@ -0,0 +1,178 @@
+        final ContextMap.Key<?> oldCmKey = acmToCm.putIfAbsent(acmKey, cmKey);
+        final AsyncContextMap.Key<?> oldAcmKey = cmToAcm.putIfAbsent(cmKey, acmKey);
+        if (oldCmKey != null || oldAcmKey != null) {
+            LOGGER.warn("Tried to register a new mapping of AsyncContextMap.Key={} to {}, but other mappings already " +
```
"AsyncContextMap.Key={} to {}" -> "AsyncContextMap.Key={} to ContextMap.Key={}".
Same below?

idelpivnitskiy replied:
No, the `ContextMap#toString()` implementations already print it consistently using a shared utility: 
https://github.com/apple/servicetalk/blob/26422543e8ce5e4b03aa65f610641ba5a5e46d7a/servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/ContextMapUtils.java#L41-L42

tkountis commented on servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/auth/BasicAuthHttpServiceFilter.java:
```diff
@@ -141,16 +148,43 @@ public Builder(final CredentialsVerifier<UserInfo> credentialsVerifier,
+         * @param userInfoAsyncContextKey a key to store a user info object in {@link AsyncContext}
+         * @return {@code this}
+         */
+        public Builder<UserInfo> userInfoAsyncContextKey(final ContextMap.Key<UserInfo> userInfoAsyncContextKey) {
```
Consider offering a `holder` / `storage` API to instruct the filter which storage to use for the user-info instead of having storage specific APIs?

idelpivnitskiy replied:
This is an interesting idea... After thinking about a `BiConsumer<UserInfo, HttpRequestMetaData>` API, decided to keep as-is for the following reasons:
1. It's easier for users to migrate from the existing deprecated method. You just change the method name and a key type and you are done. No need to learn new API.
2. Harder to explain them options how users can use this consumer and where they can store the `UserInfo`.
3. Currently, it also requires to pass `HttpRequestMetaData` to let them put info into a request context. If later we add `ContextMap` feature for a connection, we will have to change API to take `ConnectionInfo` or provide a new API that targets `ConnectionInfo`.
4. No need to handle exceptions from the user-provided function/consumer.

If you have other ideas, please share.

tkountis reviewed (approved):
AsyncContextMapToContextMapAdapter from servicetalk-concurrent-api, servicetalk-http-utils, servicetalk-http-security-jersey, servicetalk-grpc-api, servicetalk-log4j2-mdc-utils, servicetalk-opentracing-asynccontext, servicetalk-examples-http-opentracing

LGTM.

Minor comments from going through some of the other filtes.

chemicL commented on servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/AsyncContextDisableNewTest.java:
```diff
@@ -0,0 +1,116 @@
+                if (executor2 != null) {
+                    executor2.closeAsync().toFuture().get();
+                }
+                executor.closeAsync().toFuture().get();
```
I guess you'd also want to add `AsyncContext.enable();` if anything goes wrong in the outer try block.

idelpivnitskiy replied:
Not necessary, we disable it right before the `try` block. Exceptions outside this range won't affect `AsyncContext` state.

chemicL commented on servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/CopyOnWriteAsyncContextMap.java:
```diff
@@ -15,23 +15,32 @@
+import static java.util.Objects.requireNonNull;
 
 /**
  * This class provides a Copy-on-Write map behavior is and special cased for cardinality of less than 7 elements. Less
```
This javadoc line needs grammatical redaction. Not sure what the intention was but perhaps "This class provides a copy-on-write map behavior. It features special cases for cardinality of less than 7 elements." The rest seems ok, apart from "paris" -> "pairs".

chemicL reviewed (approved):
Reviewed `concurrent-api`, `concurrent-api-test` and `concurrent-api-internal` modules. LGTM.
